lib/sles4sap/database_hana.pm

```diff
@@ -0,0 +1,173 @@
+
+=over
+
+=item * B<instance_id>: Database instance ID. Mandatory.
+
+=item * B<switch_user>: Execute command as specified user with help of C<sudo su ->. Default: undef
+
+=item * B<command>: HDB command to trigger. Default: stop
+
+=back
+
+=cut
+
+sub hdb_stop {
+    my (%args) = @_;
+    my $stop_timeout = 600;
+    $args{command} //= 'stop';
+    croak("Command '$args{command}' is not supported.") unless grep(/$args{command}/, ('kill', 'stop'));
+
+    $args{command} = 'kill -x' if ($args{command} eq 'kill');
+    my $sudo_su = $args{switch_user} ? "sudo su - $args{switch_user} -c" : '';
+    my $cmd = join(' ', $sudo_su, '"', 'HDB', $args{command}, '"');
+    record_info('HDB stop', "Executing '$cmd' on " . script_output('hostname'));
+    assert_script_run($cmd, timeout => $stop_timeout);
+
+    # Wait Hana processes to stop
+    sapcontrol_process_check(instance_id => $args{instance_id}, expected_state => 'stopped', wait_for_state => 'yes', timeout => $stop_timeout);
+    record_info('DB stopped');
+}
+
+=head2 wait_for_failed_resources
+
+    wait_for_failed_resources();
+
+Wait until 'crm_mon' starts showing failed resources. This can be used as first indicator of a started failover.
+
+=cut
+
+sub wait_for_failed_resources {
+    my $timeout = 300;
+    my $start_time = time;
+    while (check_crm_output(input => script_output('crm_mon -R -r -n -N -1', quiet => 1))) {
+        sleep 30;
+        die("Cluster did not register any failed resource within $timeout sec") if (time - $timeout > $start_time);
+    }
+    record_info('CRM info', "Cluster registered failed resources\n" . script_output('crm_mon -R -r -n -N -1', quiet => 1));
+}
+
+=head2 wait_for_takeover
+
+    wait_for_takeover(target_node=>'expeliarmus');
+
+Waits until B<target_node> performs takeover and reaches 'PRIM' state.
+
+=over
+
+=item * B<target_node>: Node hostname which is expected to take over.
+
+=back
+
+=cut
+
+sub wait_for_takeover {
+    my (%args) = @_;
+    my $timeout = 300;
+    my $start_time = time;
+    my $topology;
+    my $takeover_ok;
+    until ($takeover_ok) {
+        die("Node '$args{target_node}' did not take over within $timeout sec") if (time - $timeout > $start_time);
+        $topology = calculate_hana_topology(input => script_output('SAPHanaSR-showAttr --format=script'));
+        $takeover_ok = 1 if (get_primary_node(input => $topology) eq $args{target_node});
+        sleep 30;
+    }
+}
+
+=head2 register_replica
+
+    register_replica(target_hostname=>'Dumbledore', instance_id=>'00' [, switch_user=>'hdbadm']);
+
+Executes replica node registration after failover using 'hdbnsutil' command. Node must be stopped, otherwise command fails.
+
+=over
+
+=item * B<target_hostname>: Hostname of the node that should be registered as replica
+
+=item * B<instance_id>: Instance ID
+
+=item * B<switch_user>: Execute command as specified user with help of C<sudo su ->. Default: undef
+
+
+=back
+
+=cut
+
+sub register_replica {
```

Comment:
Any reason for this new function instead of expanding `lib/sles4sap::do_hana_sr_register()`?

Reply:
I wanted to avoid bringing in `lib/sles4sap.pm` as it has requirements before imports like mandatory OpenQA variables which are not really necessary for the function.